### ISSUE TITLE
Restoring position and size after undock

### ISSUE THREAD

**Issue by PointlessReboot** [ocornut/imgui#2576]
**Version/Branch of Dear ImGui:**

Version: 1.69
Branch: docking

**Back-end/Renderer/Compiler/OS**

Back-ends: custom
Compiler: vc++ 2017
Operating System: windows

**My Issue/Question:**

I was wondering if there is anyway to keep the undocked position/size, so when you undock it can restore it. I am doing this via code, but size would be nice when draging to undock also..

I had a look through and couldn't see anything obvious. If it's not any thoughts on how it might be implemented?

Cheers

**Comment by ocornut** [comment]
This has been requested for Size before so I'd assume it is desirable.
I don't quite understand the purpose for Position since undocking is generally a manual dragging act.

Implement it would probably involve storing extra fields in `ImGuiWindow` and have them be stored in the .ini file. And/or technically speaking we might not need to store the current position/size in the .ini if we are being docked (as it will be rewritten by the docking), and we could consider just storing that undocked pos/size in the .ini file. So, save one or the other based on whether we are currently docked, and restore them when undocking.

**Comment by wolfpld** [comment]
Hey, is there any progress on this?

**Comment by ocornut** [comment]
Hello,

I have a branch with it but last time I looked the flexible docking model of dear imgui made the concept of "undocked size" very ambiguous in certain use cases. I don't remember which right now, but I've recently been pinged on this topic elsewhere so I'm going to have a look at it.

**Comment by wolfpld** [comment]
Yeah, this is not as simple as it looks. The basic expectation is that you have a window of given dimensions, which you dock somewhere, and it fills the available space. Then you undock it and the original window dimensions are restored.

Now, let's assume you have docked another window in the same place, to create a tabbed dock group. Then you undock the whole group. Which window size should be restored now? What happens when you break the dock group to have the two original windows?

**Comment by baboongd** [comment]
bump for status check (sorry)

**Comment by Nahor** [comment]
I'd like to see it too. This causes problem when undocking while in fullscreen mode, because now the window is outside the main viewport, so it's rendered in its own desktop window instead of being rendered inside the fullscreen one.

**Comment by ocornut** [comment]
You should probably disable multi-viewports when in fullscreen mode.. 

Also a few versions ago we added special code when undocking viewport-or-monitor sized window to slightly scale them down on undock to facilitate interactions.

**Comment by Nahor** [comment]
Multi-viewport is still useful when working with dual monitors:
- an outside window for statistics and other "read-only" info about the main viewport (to not cover anything in the main viewport)
- an inside-main-viewport window for interactive widgets (so the monitor doesn't need to switch between the fullscreen app and the desktop view constantly)

**Comment by trapazza** [comment]
Hello :) 

Just wondering if this is being looked at or there are plans to be looked at sometime in the near future?
I used to live with it until I got an ultrawide monitor, which magnifies the issue by a factor of 10 I'd say.

**Comment by trapazza** [comment]
Made a quick and very simple test if anyone is interested (probably I'm not taking into account some stuff)
![image](https://github.com/ocornut/imgui/assets/8068377/faffaa43-274e-40d7-bdb2-a32dc74e010f)
https://github.com/ocornut/imgui/assets/8068377/715e8f6c-3e47-47cd-82ab-d22184cc99c0

**Comment by trapazza** [comment]
Gave it a quick, dirty shot that worked, apparently:
https://github.com/ocornut/imgui/assets/8068377/de89391b-a701-4fdf-a722-faf9786d597f

If anyone's insterested, that's the simple idea behind:
![image](https://github.com/ocornut/imgui/assets/8068377/8865252a-7a4a-4ebf-8825-83c2464d78c3)

**Comment by ocornut** [comment]
Attached is a WIP patch from June 2020, rebased on latest.

[imgui-112a7a1-WIP preserve undocked size (2576).patch](https://github.com/ocornut/imgui/files/12225298/imgui-112a7a1-WIP.preserve.undocked.size.2576.patch)

It mostly appears to work at first glance but I am pretty sure there are ambiguous cases which are tricky to settle on and may cause user confusion. I'll make an effort to write them down here if I can remember them :)

**Comment by JoshuaS3** [comment]
Note on the patch: the behavior described in `imgui.cpp:4465/ImGui::StartMouseMovingWindowOrNode` ("Alter pivot to project ourselves on undocked") wasn't working for me, I had to move it to `imgui.cpp:16257/ImGui::DockNodeStartMouseMovingWindow`.

**Comment by nickvonkaenel** [comment]
I would find this feature useful as well. Docked windows in my use case are often either large or awkward shapes and if you are switching them from docked/undocked state it becomes cumbersome to always have to resize.

**Comment by Jermmyy** [comment]
I currently do this by saving the size of the window inside of `DockContextProcessDock` where i capture the `SizeFull` of the `payload_window` if it is currently floating (using` IsFloatingNode()`). Then in `DockContextProcessUndockNode` i check for a saved size and then force `node->Size` and `node->SizeRef` to those dimensions.

Here is the patch if anyone wants to take a look/give feedback

[imgui_undock_size_fix.patch](https://github.com/user-attachments/files/26750704/imgui_undock_size_fix.patch)
